### PR TITLE
Check certificate profiles and its validity period

### PR DESCRIPTION
**Pull request by shigeki**
This makes the following checks to the leaf certificate if it has
- a valid public key corresponding to its private key
- a domain in CN or SubjectAltname specified in URLSet.Sign.Domain
- a validity period within 90 days according to SXG cert requirements

### CODE REVIEW

**Comment by shigeki** [comment]
After 2019-08-01, we should cleanup this fix and `testdata/b1/*`.

**Comment by twifkak** [comment]
Nice! Despite not having a bug #, this seems pretty important for usability. I'm on vacation for the next week, so I may not have time to review until I get back. Hope that's okay.

@alin04 In case you have time to review.

File: cmd/amppkg/main.go [hunk 1]
```diff
@@ -94,13 +94,19 @@ func main() {
...
 
 	key, err := util.ParsePrivateKey(keyPem)
 	if err != nil {
 		die(errors.Wrapf(err, "parsing %s", config.KeyFile))
 	}
-	// TODO(twifkak): Verify that key matches certs[0].
+
+	for i := range config.URLSet {
+		domain := config.URLSet[i].Sign.Domain
+		err := util.CheckCertificate(certs[0], key, domain)
```
Comment (twifkak) [inline cmd/amppkg/main.go]: Optionally:

```
if err := util.CheckCertificate(certs[0], key, domain); err != nil {
  die(...)
}
```

File: packager/util/util.go [hunk 2]
```diff
@@ -76,3 +78,32 @@ func CanSignHttpExchanges(cert *x509.Certificate) bool {
...
+	if certPubKey.Y.Cmp(pubKey.Y) != 0 {
+		return errors.New("PublicKey.Y not match")
+	}
+	if err := cert.VerifyHostname(domain); err != nil {
+		return err
+	}
+	// https://wicg.github.io/webpackage/draft-yasskin-http-origin-signed-responses.html#cross-origin-cert-req
+	// Clients MUST reject certificates with this extension that were issued after 2019-05-01 and have a Validity Period longer than 90 days.
+	// After 2019-08-01, clients MUST reject all certificates with this extension that have a Validity Period longer than 90 days.
+	may1st := time.Date(2019, time.May, 1,  0, 0, 0, 0, time.UTC)
```
Comment (twifkak) [inline packager/util/util.go]: Move these to the top-level outside the function, and give them semantic names, e.g.:

```
// https://wicg.github.io/webpackage/draft-yasskin-http-origin-signed-responses.html#cross-origin-cert-req
// Clients MUST reject certificates with this extension that were issued after 2019-05-01 and have a Validity Period longer than 90 days.
// After 2019-08-01, clients MUST reject all certificates with this extension that have a Validity Period longer than 90 days.
var start90DayGracePeriod = time.Date(2019, time.May, 1,  0, 0, 0, 0, time.UTC)
var end90DayGracePeriod = time.Date(2009, time.August, 1, 0, 0, 0, 0, time.UTC)
```

File: packager/util/util_test.go [hunk 3]
```diff
@@ -27,3 +35,41 @@ func TestCanSignHttpExchanges(t *testing.T) {
...
+}
+
+func TestParse91DaysCertificate(t *testing.T) {
+	assert.Contains(t, errorFrom(util.CheckCertificate(pkgt.B3Certs91Days[0],
+		pkgt.B3Key, "amppackageexample.com")), "Certificate MUST have a Validity Period no greater than 90 days")
+}
+
+func TestParseCertificateIssuedBeforeMay1(t *testing.T) {
+	aug1st := time.Date(2009, time.August, 1, 0, 0, 0, 0, time.UTC)
+	if time.Now().After(aug1st) {
```
Comment (twifkak) [inline packager/util/util_test.go]: Is it possible to make this test stable in the face of an ever-changing `time.Now()`? I suggest one of two options:

 - Pass `now time.Date` as another param to `util.CheckCertificate`.
 - Temporarily modify `start90DayGracePeriod` and `end90DayGracePeriod` (introduced per my other suggestion), to elicit the desired test conditions.

The benefit is that both cases are tested now -- we don't have a surprise test failure on 1 Aug. (I trust that you tested the Aug 1 case, but don't trust that future changes will.)

I have a slight preference for the first option, seems simpler, but not a strong preference.

File: testdata/b3/ca.cert [hunk 4]
```diff
@@ -0,0 +1,21 @@
+-----BEGIN CERTIFICATE-----
```
Comment (twifkak) [inline testdata/b3/ca.cert]: Can you add a README.md to this directory, with instructions (e.g. shell commands) to generate these?

File: cmd/amppkg/main.go [hunk 5]
```diff
@@ -94,13 +94,19 @@ func main() {
...
 	}
-	// TODO(twifkak): Verify that certs[0] covers all the signing domains in the config.
 
 	key, err := util.ParsePrivateKey(keyPem)
 	if err != nil {
 		die(errors.Wrapf(err, "parsing %s", config.KeyFile))
 	}
-	// TODO(twifkak): Verify that key matches certs[0].
+
+	for i := range config.URLSet {
```
[the remark this answers is not in the document]
Comment (shigeki) [reply cmd/amppkg/main.go]: fixed

File: cmd/amppkg/main.go (hunk 1 above)
Comment (shigeki) [reply cmd/amppkg/main.go]: fixed

File: packager/util/util.go (hunk 2 above)
Comment (shigeki) [reply packager/util/util.go]: fixed

File: packager/util/util_test.go (hunk 3 above)
Comment (shigeki) [reply packager/util/util_test.go]: fixed with the first option.

File: testdata/b3/ca.cert (hunk 4 above)
Comment (shigeki) [reply testdata/b3/ca.cert]: Added

**Comment by shigeki** [comment]
Thanks for reviewing. Fixed with your comments. PTAL.

**Pull request merged**